File: CSharp/Dia01.cs
using System;
using System.Collections.Generic;

namespace CSharp
{
    public class Dia01
    {
        private static (int, int)? Find2(int valor, IReadOnlyList<int> numeros)
        {
            var posActual = 0;
            while (posActual < numeros.Count)
            {
                var elemActual = numeros[posActual];
                for (var i = posActual + 1; i < numeros.Count; i++)
                {
                    if (elemActual + numeros[i] == valor)
                    {
                        return (elemActual, numeros[i]);
                    }
                }

                posActual += 1;
            }

            return null;
        }

        public void Puzzle1(string ruta)
        {
            var lineas = System.IO.File.ReadAllLines(ruta);
            var numeros = new int[lineas.Length];
            for (var i = 0; i < lineas.Length; i++)
            {
                numeros[i] = int.Parse(lineas[i]);
            }

            var resultado = Find2(2020, numeros);
            if (resultado != null)
            {
                var v1 = resultado.Value.Item1;
                var v2 = resultado.Value.Item2;
                Console.WriteLine("Números: {0}, {1}, {2}", v1, v2, v1 * v2);
            }
            else
            {
                Console.WriteLine("Números no encontrados :c");
            }
        }
    }
}

File: elixir/lib/dia01.ex
defmodule Advent.Dia01 do
  @moduledoc false

  defp find2p(value, numbers) do
    pidActual = self()
    numElems = Enum.count(numbers)
    indices = 0..numElems
    Enum.each(indices, fn i ->
      spawn fn ->
        item = Enum.at(numbers, i)
        nuevosIndices = (i + 1)..numElems

        Enum.each(nuevosIndices, fn j ->
          case Enum.at(numbers, j, nil) do
            nil -> 0
            nItem ->
              if item + nItem == value do
                send(pidActual, {:ok, item, nItem})
              end
          end
        end)

      end
    end)

    receive do
      {:ok, v1, v2} -> IO.puts "Encontrado con procesos: #{v1}, #{v2} -> #{v1 * v2}"
    end
  end

  defp find2(value, numbers) do
    case numbers do
      [x | xs] ->
        case Enum.find(xs, fn v -> x + v == value end) do
          nil -> find2(2020, xs)
          v -> {:ok, x, v}
        end
      [] -> {:err}
    end
  end

  def puzzle1(rutaArchivo) do
    {:ok, str} = File.read(rutaArchivo)
    numeros = Enum.map(String.split(str, "\n"), &(if &1 == "" do 0 else String.to_integer(&1) end))

    # find2p(2020, numeros)

    t1 = :os.system_time(:microsecond)
    strRes =
      case find2(2020, numeros) do
        {:ok, v1, v2} -> "El resultado es #{v1 * v2}"
        {:err} -> "Numeros no encontrados."
      end
    t2 = :os.system_time(:microsecond)
    IO.puts "#{t2 - t1}"
    IO.puts "Dia 01 puzzle 1: #{strRes}"
  end

end
